README.md
### Skydock - Automagic Service Discovery for [Docker](https://github.com/dotcloud/docker)


## BEFORE GO
The main code is in directory ```src/github.com/crosbymichael/skydock```


## NOTICE

I plan on making some breaking changes soon to help skydns and skydock scale better.  To stay up-to-date either
watch this repo or follow me on twitter @crosbymichael.  


Skydock monitors docker events when containers start, stop, die, kill, etc and inserts records into a dynamic
DNS server [skydns](https://github.com/skynetservices/skydns1).  This allows standard DNS queries for services
running inside docker containers.  Because lets face it, if you have to modify your application code to work
with other service discovery solutions you might as well just give up.  DNS just works and it works well.  
Also you cannot be expected to modify application code that you don't own.  Passing service urls via the
cli or in static config files (nginx) will not be possible if your service discovery solution requires
a client library just to fetch an IP.  


[Skydns](https://github.com/skynetservices/skydns1) is a very small and simple server that does DNS for 
discovery very well.  The authors and contributors to skydns helped a lot to make this project possible.
Skydns exposes a very simple REST API to add, update, and remove services.


#### The Details

When you start a container with docker an event is sent to skydock via the events endpoint.  Skydock will
then inspect the container's information and add an entry to skydns.  We setup skydns to bind it's nameserver
to the **docker0** bridge so that it is available to containers.  For DNS queries that are not part of the 
domain registered with skydns for service discovery, skydns will forward the query to an authoritative nameserver.
Skydns will return A, AAAA, and SRV records for registered services.



When designing skydock I made the assumption that when in the context of service discovery a client does
not care about a specific instance of a service running inside a container.  The client cares about the 
service and in the context of docker a service is defined by an image.  We have many images; redis, postgres,
our frontend applications, queues, and workers.  The URL scheme is designed using this assumption.


**Parts of the URL**
* Domain (domain name to resolve DNS requests for)
* Environment (context of what type of service is running dev, production, qa, uat)
* Service (the actual service name derived from the image name minus the repository crosbymichael/redis -> redis)
* Instance (container's name representing the actual instance of a service)
* Region (currently not used but will be your docker host, digitalocean, ec2; this will be used for multihost)


A typical query will look like this if your domain is `crosbymichael.com` and environment is `production`:

```bash
curl webapp.production.crosbymichael.com
```

The query above will return the IP for a container running the image webapp.  If we want a specific instance
we can prepend the container name to the query.  If our webapp container's name was webapp1 we could do this 
to get the specific container.

```bash
curl webapp1.webapp.production.crosbymichael.com
```

Very simple and easy and no client code was harmed in this demonstration.  Wildcard queries are also supported.  

```bash
dig @172.17.42.1 "webapp.*.crosbymichael.com"
```


#### Setup

So what type of hacks do you have to do to get this running?  Nothing, everything runs inside docker containers.
You are just two `docker run` s away from bliss.


Ok, I lied.  You have to make one change to your docker daemon.  You need to run the daemon with the `-dns` flag so
that docker injects a specific nameserver into the `/etc/resolv.conf` of each container that is run.  First get the 
IP address of the `docker0` bridge.  We will need to know the IP of the bridge so that we can tell skydns to bind it's 
nameserver to that IP.  For this example we will use the ip `172.17.42.1` as the `docker0` bridge.


```bash
# start your daemon with the -dns flag, figure it out...
docker -d --bip=172.17.42.1/16 --dns=172.17.42.1 # + what other settings you use
```

**Note:**
You can also pass the `-dns` flag to individual containers so that the DNS options only apply to specific 
containers and not everything started by the daemon.  But what fun is that?

Now we need to start skydns before our other containers are run or else they will not be able to resolve DNS queries.

```bash
docker pull crosbymichael/skydns
docker run -d -p 172.17.42.1:53:53/udp --name skydns crosbymichael/skydns -nameserver 8.8.8.8:53 -domain docker
```

We add the name skydns to the container and we use `-p` and tell docker to bind skydns port 53/udp to the docker0 bridge's IP.
We give skydns a nameserver of `8.8.8.8:53`.  This nameserver is used to forward queries that are not for service discovery to a 
real nameserver.  If you don't know `8.8.8.8` is google's public DNS address.


Next is the `-domain` flag.  This is the domain that you want skydns to resolve DNS queries for.  In this example I am running 
docker on my local development machine so I am using the domain name `docker`.  Any requests for services `*.docker` will be 
resolved by skydns for service discovery, all other requests will be forwarded to `8.8.8.8`.


Now that skydns is running we can start skydock to bridge the gap between docker and skydns.


```bash
docker pull crosbymichael/skydock
docker run -d -v /var/run/docker.sock:/docker.sock --name skydock crosbymichael/skydock -ttl 30 -environment dev -s /docker.sock -domain docker -name skydns
```


This one is as little more involved but the parts are still simple.  First we give it the name of skydock and we bind docker's unix socket into the container.
I'm guessing for most, you do not want to service the docker API on a tcp port for containers to reach.  If we bind the unix socket into this container we don't 
have to worry about other containers accessing the API, only skydock.  We also add a link for skydns so that skydock knows where to make requests to insert 
new records.  We are pre DNS discovery at this point.


Now we have a few settings to assign to skydock.  First is the TTL value that you want all services to have when skydock adds them to DNS.  I'm using a
TTL value 30 seconds but you can set it higher or lower if needed.  Skydock will also start a heartbeat for the service after it is added.  You can use 
the `-beat` flag to set this default interval in seconds for the heartbeat or skydock will set the heartbeat interval to `TTL -(TTL/4)`.  I know, too 
complicated.


Next is the `-environment` flag which is the second part of your DNS queries.  I set this to `dev` because it is running on my local machine.  `-s` is 
the final option and it just tells skydock where to find docker's unix socket so that it can make requests to docker's API.


Now you're done.  Just start containers and use intuitive urls to discover your services.  Here is an small example starting a redis server and connecting 
the redis-cli to that instance of the service.  Because it's DNS you can specific the urls on `docker run`.  


```bash
# run an instance of redis
docker run -d --name redis1 crosbymichael/redis
03582c0de0ebb10665678d6ed530ae98bebd7d63dad5e7fb1cd53ffb1f85d91d

# run the cli and connect to our new instance
docker run -t -i crosbymichael/redis-cli -h redis1.redis.dev.docker

redis.dev.docker:6379> set name koye
OK
redis.dev.docker:6379> get name
"koye"
redis.dev.docker:6379>

```

That is, the `redis1` named `crosbymichael/redis` container was available under the hostname `redis1.redis.dev.docker`.

```
dig @172.17.42.1 +short redis1.redis.dev.docker
172.17.0.4
```

If you were to run additional `crosbymichael/redis` containers, they would all be available under the `redis.dev.docker` hostname.

```
docker run -d --name redis2 crosbymichael/redis
docker run -d --name redis3 crosbymichael/redis

dig @172.17.42.1 +short redis.dev.docker
172.17.0.4
172.17.0.5
172.17.0.6
```

#### Plugin support
I just added plugin support via [otto](https://github.com/robertkrimen/otto) to allow users to write plugins in javascript.  Currently only one function uses plugins and that is `createService(container)`.  This function takes a container's configuration and converts it into a DNS service  The current functionality is implementing in this javascript function:

```javascript
function createService(container) {
    return {
        Port: 80,
        Environment: defaultEnvironment,
        TTL: defaultTTL,
        Service: cleanImageName(container.Image),
        Instance: removeSlash(container.Name),
        Host: container.NetworkSettings.IpAddress
    }; 
}
```

Your function must be called `createservice` which takes one object, the container, and must return a service with the fields shown above.  In your plugin you have access to the following global variables and functions.


```javascript
var defaultEnvironment = "string - the environment from the -environment flag";
var defaultTTL = 30; // int - the ttl value from the -ttl flag

function cleanImageName(string) string // cleans the repo and tags of the passed parameter returning the result
function removeSlash(string) string  // removes all / from the passed parameter returning the result
```

And that is it.  Just add a `createservice` function to a .js file then use the `-plugins` flag to enable your new plugin.  Plugins are loaded at start so changes made to the functions during the life of skydock are not reflected, you have to restart ( done for performance ).  

```bash
docker run -d -v /var/run/docker.sock:/docker.sock -v /myplugins.js:/myplugins.js --name skydock --link skydns:skydns crosbymichael/skydock -s /docker.sock -domain docker -plugins /myplugins.js
```

Feel free to submit your plugins to this repo under the `plugins/` directory.  


#### TODO/ROADMAP
* Multihost support
* Handle multiple ports via SRV records

#### Bugs
* Please report all skydock bugs on this repository
* Report all skydns bugs [here](https://github.com/skynetservices/skydns1/issues?state=open)

#### License - MIT

Copyright (c) 2014 Michael Crosby. michael@crosbymichael.com

Permission is hereby granted, free of charge, to any person
obtaining a copy of this software and associated documentation 
files (the "Software"), to deal in the Software without 
restriction, including without limitation the rights to use, copy, 
modify, merge, publish, distribute, sublicense, and/or sell copies 
of the Software, and to permit persons to whom the Software is 
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be 
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED,
INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, 
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. 
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT 
HOLDERS BE LIABLE FOR ANY CLAIM, 
DAMAGES OR OTHER LIABILITY, 
WHETHER IN AN ACTION OF CONTRACT, 
TORT OR OTHERWISE, 
ARISING FROM, OUT OF OR IN CONNECTION WITH 
THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
# Release notes for project skydock


CHANGELOG
---------

### 1.0.0


* Final version working
* Almost working
* Skydock works again
* importing skydocker and dependencies source code
* refs &#35;99; move import names back; gogmt;
* refs &#35;99; fix issue when migrating to docker 1.11 (docker API 1.23); contaner State is now string;
* skydns repository is now skydns2, old code is at skydns1
* Update influxdb to new repository
* Use --bip setting to configure bridge IP.
* Update readme with notice
* Use direct container name lookup instead of link
* Update dockerfile to be selfsufficent
* change deprecated -link to --link
* Switch to "debian:jessie" as the Dockerfile base
* [readme] Further investigation into hostnames
* Fixed use of docker run ... --name
* Example Nginx configuration, resolves services at runtime.
* Use uuid instead of container name
* Add function to find exposed ports
* Move default plugin to js file
* Fix signal handling on kill
* Fix utils tests
* Update travis script for otto
* Add env var plugin as example
* Add plugin support via javascript
* Make sure to unlock if container exists
* Using the utils package correctly for SplitURI
* Added support for passing path only to SplitURI
* Removed old conflicting files after rebase and added SplitURI tests
* Added support for passing a URI (tcp:// http:// unix:// etc.) for the -s option
* Don't hit docker for heartbeats
* Handle signal interrupt to close socket
* Fix events handlers
* parse private registry image names
* reset ttl on restore conflicts
* Update logging
* Update travis deps
* Ignore added services with conflict and start heartbeat
* Add leveled logger
* Update skydns client for changes
* Add travis badge to readme
* Add travis file
* Add testing for main functions
* Move docker requests into docker pkg
* Move utils to sub pkg and add unit tests
* Add handlers in goroutines to handle concurrent events
* small typo
* Fix typo in readme for domain
* Add domain to readme
* Update to not use the same client conn
* Continue after decoding error
* Add docker client to remove boiler plate
* Remove run script
* Update wildcard curl example to use dig
* Fix not tagged image on heartbeat
* Add not about -dns per container
* Skip untagged images
* Refactoring of code
* Change hostname to domain
* Add readme
* go fmt
* refactoring
* removed duplicate func for truncate
* restore containers on start
* Add license file
* Update for new skydns client pkg
* Add dockerfile
* Add support for reading from the unix socket
* Improve dns query layout and env
* Add heartbeat for services
* Fix uuid on remove
* Initial commit


________

Generated by sr-release. 


test.sh
#!/usr/bin/env bash


export ROOT_URL_PREFIX=lmac
export ROOT_URL_DOMAIN=linguahouse.com
docker stop skydock
docker rm skydock
#docker run --restart=always -d -v /var/run/docker.sock:/docker.sock --name skydock -l com.mobibeam.docker.name=skydock docker.mobibeam.com/skydock -ttl 600 -environment "$ROOT_URL_PREFIX" -s "/docker.sock" -domain "$ROOT_URL_DOMAIN" -skydns "http://skydns.mobibeam.com:8080" -secret alamakaca -plugins /plugins/rancher.js

docker run --restart=always -d -v /var/run/docker.sock:/docker.sock --name skydock  -l  "io.rancher.container.ip=10.42.246.232/16" docker.mobibeam.com/skydock -ttl 60 -beat 10 -environment "$ROOT_URL_PREFIX" -s "/docker.sock" -domain "$ROOT_URL_DOMAIN" -skydns "http://skydns00.linguahouse.com:8080"  -secret alamakaca

plugins/rancher.js
// this plugin is seeting names based on the rancher labels

function objdump(arr, level) {
    var dumped_text = "";
    if(!level) level = 0;

    var level_padding = "";
    for(var j=0;j<level+1;j++) level_padding += "    ";

    if(typeof(arr) == 'object') {
        for(var item in arr) {
            var value = arr[item];

            if(typeof(value) == 'object') {
                dumped_text += level_padding + "'" + item + "' ...\n";
                dumped_text += objdump(value,level+1);
            } else {
                dumped_text += level_padding + "'" + item + "' => \"" + value + "\"\n";
            }
        }
    } else {
        dumped_text = "===>"+arr+"<===("+typeof(arr)+")";
    }
    return dumped_text;
}


function createService(container) {
    var port = getDefaultPort(container);
    var address = container.NetworkSettings.IpAddress;
    
    console.log("Rancher plugin in use");

    if (container.Config.Labels['io.rancher.container.ip']){
        var rancherAddress=container.Config.Labels['io.rancher.container.ip'];
        console.log("Detected rancher address: "+rancherAddress);
        var slashpos = rancherAddress.indexOf("/");
        if (slashpos){
            address=rancherAddress.substring(0,slashpos);
        }else{
            address=rancherAddress;
        }
        console.log("Address to be used is "+address);
    }
    
    return {
        Port: port,
        Environment: defaultEnvironment,
        TTL: defaultTTL,
        Service: cleanImageName(container.Image),
        Instance: removeSlash(container.Name),
        Host: address
    }; 
}

function getDefaultPort(container) {
    // if we have any exposed ports use those
    var port = 0;
    var ports = container.NetworkSettings.Ports;
    if (Object.keys(ports).length > 0) {
        for (var key in ports) {
            var value = ports[key];
            if (value !== null && value.length > 0) {
                for (var i = 0; i < value.length; i++) {
                    var hp = parseInt(value[i].HostPort);
                    if (port === 0 || hp < port) {
                        port = hp; 
                    }
                }  
            } else if (port === 0) {
                // just grab the key value 
                var expose = parseInt(key.split("/")[0]); 
                port = expose;
            }
        }
    } 
     
    if (port === 0) {
        port = 80; 
    }
    return port;
}


src/github.com/crosbymichael/log/log.go
package log

import (
	"fmt"
)

const (
	// FATAL error log levels
	FATAL = iota // fatal errors

	// ERROR error log levels
	ERROR = iota // errors might happend

	// INFO log only info level
	INFO = iota // debug mode

	// DEBUG level
	DEBUG = iota // debug mode
)

// Logger is a struct ..
type Logger struct {
	level int
	name  string
}

var currentlog Logger

// Logf is a logging function
func Logf(level int, format string, v ...interface{}) {
	if currentlog.level <= level {
		if v != nil {
			fmt.Printf(format, v...)
			fmt.Println("")
		} else {
			fmt.Println(format)
		}

	}
}

// SetLogger defines logger pameters in the current conext
func SetLogger(logger Logger) error {
	// currentlog := logger
	return nil
}

// NewStandardLevelLogger creates a new statand logger
func NewStandardLevelLogger(n string) Logger {
	logger := Logger{level: INFO, name: n}
	return logger
}


src/github.com/crosbymichael/skydock/docker/client.go
package docker

import (
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net"
	"net/http"
	"net/http/httputil"
	"os"
	"os/signal"
	"syscall"

	"github.com/crosbymichael/log"
	"github.com/crosbymichael/skydock/utils"
)

type (
	Docker interface {
		FetchAllContainers() ([]*Container, error)
		FetchContainer(name, image string) (*Container, error)
		GetEvents() chan *Event
	}

	Event struct {
		ContainerId string `json:"id"`
		Status      string `json:"status"`
		Image       string `json:"from"`
	}

	ContainerConfig struct {
		Hostname string
		Domainname string
		User string
		Tty bool
		Env      []string
		Cmd []string
		Image    string
		Entrypoint []string
		MacAddress string
		Labels map[string]string
	}

	Binding struct {
		HostIp   string
		HostPort string
	}

	NetworkSettings struct {
		IpAddress string
		Ports     map[string][]Binding
	}

	// GET /containers/json returns the state of the container, one of:
	// - created;
	// - restarting;
	// - running;
	// - paused;
	// - exitedor;
	// - dead;
	State string

	Container struct {
		Id              string
		Image           string
		Name            string
		Config          *ContainerConfig
		NetworkSettings *NetworkSettings
		//State           State
	}

	dockerClient struct {
		path string
	}
)

var (
	ErrImageNotTagged = errors.New("image not tagged")
)

func NewClient(path string) (Docker, error) {
	return &dockerClient{path}, nil
}

func (d *dockerClient) newConn() (*httputil.ClientConn, error) {
	prot, path := utils.SplitURI(d.path)
	conn, err := net.Dial(prot, path)
	if err != nil {
		return nil, err
	}
	return httputil.NewClientConn(conn, nil), nil
}

func (d *dockerClient) FetchContainer(name, image string) (*Container, error) {
	c, err := d.newConn()
	if err != nil {
		return nil, err
	}
	defer c.Close()

	req, err := http.NewRequest("GET", fmt.Sprintf("/containers/%s/json", name), nil)
	if err != nil {
		return nil, err
	}

	resp, err := c.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusOK {
		var (
			container *Container
			d         = json.NewDecoder(resp.Body)
		)

		if err = d.Decode(&container); err != nil {
			return nil, err
		}

		// These should match or else it's from an image that is not tagged
		if image != "" && utils.RemoveTag(image) != utils.RemoveTag(container.Config.Image) {
			return nil, ErrImageNotTagged
		}
		container.Image = image

		return container, nil
	}
	return nil, fmt.Errorf("Could not fetch container %d", resp.StatusCode)
}

func (d *dockerClient) FetchAllContainers() ([]*Container, error) {
	req, err := http.NewRequest("GET", fmt.Sprintf("/containers/json"), nil)
	if err != nil {
		return nil, err
	}

	c, err := d.newConn()
	if err != nil {
		return nil, err
	}
	defer c.Close()

	resp, err := c.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode == http.StatusOK {
		var containers []*Container
		if err = json.NewDecoder(resp.Body).Decode(&containers); err != nil {
			return nil, err
		}
		return containers, nil
	}
	return nil, fmt.Errorf("invalid HTTP request %d %s", resp.StatusCode, resp.Status)
}

func (d *dockerClient) GetEvents() chan *Event {
	eventChan := make(chan *Event, 100) // 100 event buffer
	go func() {
		defer close(eventChan)

		c, err := d.newConn()
		if err != nil {
			log.Logf(log.FATAL, "cannot connect to docker: %s", err)
			return
		}
		defer c.Close()

		req, err := http.NewRequest("GET", "/events", nil)
		if err != nil {
			log.Logf(log.ERROR, "bad request for events: %s", err)
			return
		}

		resp, err := c.Do(req)
		if err != nil {
			log.Logf(log.FATAL, "cannot connect to events endpoint: %s", err)
			return
		}
		defer resp.Body.Close()

		// handle signals to stop the socket
		sigChan := make(chan os.Signal, 1)
		signal.Notify(sigChan, os.Interrupt, syscall.SIGTERM, syscall.SIGQUIT)
		go func() {
			for sig := range sigChan {
				log.Logf(log.INFO, "received signal '%v', exiting", sig)

				c.Close()
				close(eventChan)
				os.Exit(0)
			}
		}()

		dec := json.NewDecoder(resp.Body)
		for {
			var event *Event
			if err := dec.Decode(&event); err != nil {
				if err == io.EOF {
					break
				}
				log.Logf(log.ERROR, "cannot decode json: %s", err)
				continue
			}
			eventChan <- event
		}
		log.Logf(log.DEBUG, "closing event channel")
	}()
	return eventChan
}


src/github.com/crosbymichael/skydock/main.go
/*
   Multihost
   Multiple ports
*/

package main

import (
	"flag"
	"fmt"
	"os"
	"sync"
	"time"

	"github.com/crosbymichael/log"
	"github.com/crosbymichael/skydock/docker"
	"github.com/crosbymichael/skydock/utils"
	// influxdb "github.com/influxdb/influxdb/client"
	"github.com/skynetservices/skydns1/client"
	"github.com/skynetservices/skydns1/msg"
)

var (
	pathToSocket        string
	domain              string
	environment         string
	skydnsURL string
	skydnsContainerName string
	secret              string
	ttl                 int
	beat                int
	numberOfHandlers    int
	pluginFile          string

	skydns       Skydns
	dockerClient docker.Docker
	plugins      *pluginRuntime
	running      = make(map[string]struct{})
	runningLock  = sync.Mutex{}
)

func initFunc() {
	flag.StringVar(&pathToSocket, "s", "/var/run/docker.sock", "path to the docker unix socket")
	flag.StringVar(&skydnsURL, "skydns", "", "url to the skydns url")
	flag.StringVar(&skydnsContainerName, "name", "", "name of skydns container")
	flag.StringVar(&secret, "secret", "", "skydns secret")
	flag.StringVar(&domain, "domain", "", "same domain passed to skydns")
	flag.StringVar(&environment, "environment", "dev", "environment name where service is running")
	flag.IntVar(&ttl, "ttl", 60, "default ttl to use when registering a service")
	flag.IntVar(&beat, "beat", 0, "heartbeat interval")
	flag.IntVar(&numberOfHandlers, "workers", 3, "number of concurrent workers")
	flag.StringVar(&pluginFile, "plugins", "/plugins/rancher.js", "file containing javascript plugins (plugins.js)")

	flag.Parse()

}

func validateSettings() {
	if beat < 1 {
		beat = ttl - (ttl / 4)
	}

	if (skydnsURL != "") && (skydnsContainerName != "") {
		fatal(fmt.Errorf("specify 'name' or 'skydns', not both"))
	}

	if (skydnsURL == "") && (skydnsContainerName == "") {
		skydnsURL = "http://" + os.Getenv("SKYDNS_PORT_8080_TCP_ADDR") + ":8080"
	}

	if domain == "" {
		fatal(fmt.Errorf("Must specify your skydns domain"))
	}
}


func printCurrentSettings(){
	fmt.Println("Working with the fallowing parameters:")
	fmt.Printf("skydnsURL %s\n", skydnsURL)
	fmt.Printf("pathToSocket %s\n", pathToSocket)
	fmt.Printf("skydnsContainerName %s\n", skydnsContainerName)
	fmt.Printf("secret %s\n", secret)
	fmt.Printf("domain %s\n", domain)
	fmt.Printf("environment %s\n", environment)
	fmt.Printf("ttl %d\n", ttl)
	fmt.Printf("beat %d\n", beat)
	fmt.Printf("numberOfHandlers %d\n", numberOfHandlers)
	fmt.Printf("pluginFile %s\n", pluginFile)
}

func setupLogger() error {
	var (
		logger log.Logger
	)

	if host := os.Getenv("INFLUXDB_HOST"); host != "" {
		// config := &influxdb.ClientConfig{
		// 	Host:     host,
		// 	Database: os.Getenv("INFLUXDB_DATABASE"),
		// 	Username: os.Getenv("INFLUXDB_USER"),
		// 	Password: os.Getenv("INFLUXDB_PASSWORD"),
		// }
		//
		// logger, err = log.NewInfluxdbLogger(fmt.Sprintf("%s.%s", environment, domain), "skydock", config)
		// if err != nil {
		// 	return err
		// }
	} else {
		logger = log.NewStandardLevelLogger("skydock")
	}

	if err := log.SetLogger(logger); err != nil {
		return err
	}
	return nil
}

func heartbeat(uuid string, service *msg.Service) {
	runningLock.Lock()
	if _, exists := running[uuid]; exists {
		runningLock.Unlock()
		return
	}
	running[uuid] = struct{}{}
	runningLock.Unlock()

	defer func() {
		runningLock.Lock()
		delete(running, uuid)
		runningLock.Unlock()
	}()

	//var errorCount int
	for _ = range time.Tick(time.Duration(beat) * time.Second) {
		// checking if the service exists after updating
		_ , err2 := skydns.Get(uuid)
		if err2 != nil {
			log.Logf(log.ERROR, "The expected service is not registered in skydns. Trying to register service again: adding %s (%s) to skydns", uuid, service.Name)
			if err := skydns.Add(uuid, service); err != nil {
				log.Logf(log.ERROR, "Service (%s) %s  have not been addedd sucessfully. Error ", uuid, service.Name, err)
			}
		}

		err := updateService(uuid, ttl)
		if beat >=30 {
			if err != nil {
				log.Logf(log.INFO, "updating ttl for (%s) %s  fialed. Err: %s", uuid, service.Name, err)
			} else {
				log.Logf(log.INFO, "updating ttl for (%s) %s succeed", uuid, service.Name)
			}
		}
	}
}

// restoreContainers loads all running containers and inserts
// them into skydns when skydock starts
func restoreContainers() error {
	containers, err := dockerClient.FetchAllContainers()
	if err != nil {
		return err
	}

	var container *docker.Container
	for _, cnt := range containers {
		uuid := utils.Truncate(cnt.Id)
		if container, err = dockerClient.FetchContainer(uuid, cnt.Image); err != nil {
			if err != docker.ErrImageNotTagged {
				log.Logf(log.ERROR, "failed to fetch %s on restore: %s", cnt.Id, err)
			}
			continue
		}

		service, err := plugins.createService(container)
		if err != nil {
			// doing a fatal here because we cannot do much if the plugins
			// return an invalid service or error
			fatal(err)
		}
		if err := sendService(uuid, service); err != nil {
			log.Logf(log.ERROR, "failed to send %s to skydns on restore: %s", uuid, err)
		}
	}
	return nil
}

// sendService sends the uuid and service data to skydns
func sendService(uuid string, service *msg.Service) error {
	fmt.Printf("service.Name %s\n", service.Name)
	fmt.Printf("uuid %s\n", uuid)
	log.Logf(log.INFO, "adding %s (%s) to skydns", uuid, service.Name)
	err := skydns.Add(uuid, service);
	// ignore erros for conflicting uuids and start the heartbeat again
	if err == client.ErrConflictingUUID {
		log.Logf(log.INFO, "service already exists for %s. Resetting ttl.", uuid)
		updateService(uuid, ttl)
	}
	go heartbeat(uuid, service)
	return err
}

func removeService(uuid string) error {
	log.Logf(log.INFO, "removing %s from skydns", uuid)
	return skydns.Delete(uuid)
}

func addService(uuid, image string) error {
	container, err := dockerClient.FetchContainer(uuid, image)
	if err != nil {
		if err != docker.ErrImageNotTagged {
			return err
		}
		return nil
	}

	service, err := plugins.createService(container)
	if err != nil {
		// doing a fatal here because we cannot do much if the plugins
		// return an invalid service or error
		fatal(err)
	}

	if err := sendService(uuid, service); err != nil {
		return err
	}
	return nil
}

func updateService(uuid string, ttl int) error {
	return skydns.Update(uuid, uint32(ttl))
}

func eventHandler(c chan *docker.Event, group *sync.WaitGroup) {
	defer group.Done()

	for event := range c {
		log.Logf(log.DEBUG, "received event (%s) %s %s", event.Status, event.ContainerId, event.Image)
		uuid := utils.Truncate(event.ContainerId)

		switch event.Status {
		case "die", "stop", "kill":
			if err := removeService(uuid); err != nil {
				log.Logf(log.ERROR, "error removing %s from skydns: %s", uuid, err)
			}
		case "start", "restart":
			if err := addService(uuid, event.Image); err != nil {
				log.Logf(log.ERROR, "error adding %s to skydns: %s", uuid, err)
			}
		}
	}
}

func fatal(err error) {
	fmt.Fprintf(os.Stderr, "%s\n", err)
	os.Exit(1)

}

func main() {
	initFunc()
	validateSettings()
	printCurrentSettings()
	if err := setupLogger(); err != nil {
		fatal(err)
	}

	var (
		err   error
		group = &sync.WaitGroup{}
	)

	plugins, err = newRuntime(pluginFile)
	if err != nil {
		fatal(err)
	}

	if dockerClient, err = docker.NewClient(pathToSocket); err != nil {
		log.Logf(log.FATAL, "error connecting to docker: %s", err)
		fatal(err)
	}

	if skydnsContainerName != "" {
		container, err := dockerClient.FetchContainer(skydnsContainerName, "")
		if err != nil {
			log.Logf(log.FATAL, "error retrieving skydns container '%s': %s", skydnsContainerName, err)
			fatal(err)
		}

		skydnsURL = "http://" + container.NetworkSettings.IpAddress + ":8080"
	}

	log.Logf(log.INFO, "skydns URL: %s", skydnsURL)

	if skydns, err = client.NewClient(skydnsURL, secret, domain, "172.17.42.1:53"); err != nil {
		log.Logf(log.FATAL, "error connecting to skydns: %s", err)
		fatal(err)
	}

	log.Logf(log.DEBUG, "starting restore of containers")
	if err := restoreContainers(); err != nil {
		log.Logf(log.FATAL, "error restoring containers: %s", err)
		fatal(err)
	}

	events := dockerClient.GetEvents()

	group.Add(numberOfHandlers)
	// Start event handlers
	for i := 0; i < numberOfHandlers; i++ {
		go eventHandler(events, group)
	}

	log.Logf(log.DEBUG, "starting main process")
	group.Wait()
	log.Logf(log.DEBUG, "stopping cleanly via EOF")
}
